classes/surgeon.php

```diff
@@ -0,0 +1,520 @@
+            \core_tag_tag::remove_all_item_tags('core', 'course_modules', $cm->id);
+            $outcomemessages[] = get_string('outcome_tag_table_record_deleted', 'tool_fix_delete_modules');
+        }
+
+        // Notify the competency subsystem.
+        \core_competency\api::hook_course_module_deleted($cm);
+
+        // Delete the context.
+        if ($modcontext) {
+            \context_helper::delete_instance(CONTEXT_MODULE, $cm->id);
+
+            $outcomemessages[] = get_string('outcome_context_table_record_deleted', 'tool_fix_delete_modules');
+        }
+
+        // Delete the module from the course_modules table.
+        if ($DB->delete_records('course_modules', array('id' => $cm->id))) {
+            $outcomemessages[] = get_string('outcome_course_module_table_record_deleted', 'tool_fix_delete_modules');
+        } else {
+            $outcomemessages[] = get_string('outcome_context_table_record_delete_fail', 'tool_fix_delete_modules');
+        }
+
+        // Delete module from that section.
+        if (!delete_mod_from_section($cm->id, $cm->section)) {
+            $outcomemessages[] = get_string('outcome_course_section_data_deleted', 'tool_fix_delete_modules');
+        } else {
+            $outcomemessages[] = get_string('outcome_course_section_data_delete_fail', 'tool_fix_delete_modules');
+        }
+
+        // Trigger event for course module delete action.
+        if ($modcontext && $modulename) {
+            $event = \core\event\course_module_deleted::create(array(
+                'courseid' => $cm->course,
+                'context'  => $modcontext,
+                'objectid' => $cm->id,
+                'other'    => array(
+                    'modulename'   => $modulename,
+                    'instanceid'   => $cm->instance,
+                )
+            ));
+            $event->add_record_snapshot('course_modules', $cm);
+            $event->trigger();
+            // Function for Moodle 4.0+.
+            if (method_exists('\course_modinfo', 'purge_course_module_cache')) {
+                \course_modinfo::purge_course_module_cache($cm->course, $cm->id);
+            }
+            rebuild_course_cache($cm->course, true);
+        }
+
+        // Reset adhoc task to run asap.
+        $now = time();
+        if ($thisadhoctask = $this->get_adhoctask_from_taskid($task->taskid)) {
+            $thisadhoctask->set_fail_delay(0);
+            $thisadhoctask->set_next_run_time($now);
+            // Function exists on Moodle 3.7+.
+            if (method_exists('\core\task\manager', 'reschedule_or_queue_adhoc_task')) {
+                \core\task\manager::reschedule_or_queue_adhoc_task($thisadhoctask);
+            } else {
+                $this->reschedule_or_queue_adhoc_task($thisadhoctask);
+            }
+            $outcomemessages[] = get_string('outcome_adhoc_task_record_rescheduled', 'tool_fix_delete_modules');
+            try {
```

Comment:
I'm pretty sure we don't need to run the task in user session. We've rescheduled the task, re-set it faildelay and nextrun time and this is enough. Leave the rest to cron. As long as faildelay is re-set, this task should be picked up by the plugin report, so we don't need to worry about the case when someone push the button twice for one task